sklearn/preprocessing/_encoders.py

```diff
@@ -722,14 +749,39 @@ def get_feature_names_out(self, input_features=None):
         cats = self.categories_
         input_features = _check_feature_names_in(self, input_features)
 
+        name_combiner = self._get_feature_name_combiner()
         feature_names = []
         for i in range(len(cats)):
-            names = [input_features[i] + "_" + str(t) for t in cats[i]]
+            names = [name_combiner(input_features[i], t) for t in cats[i]]
             if self.drop_idx_ is not None and self.drop_idx_[i] is not None:
                 names.pop(self.drop_idx_[i])
             feature_names.extend(names)
+        feature_names = self._check_feature_names_uniqueness(feature_names)
```

Comment:
I do not think we current enforce uniqueness. For example `StandardScalar` is basically a noop returning the input features without checking for uniqueness. We also add a little more overhead for uniqueness checking.

In this case, `OneHotEncoder` will not have unique names for the edge case of `[None, "None"]`. OneHotEncoder does not usually accept multiple dtypes:

```python
from sklearn.preprocessing import OneHotEncoder

enc = OneHotEncoder()

# Fails
enc.fit_transform([[1], ["1"]])
```

I think we can reduce the scope of this PR by only including the callable support. We can follow up with a PR on uniqueness checking.

Reply:
> I do not think we current enforce uniqueness. For example StandardScalar is basically a noop returning the input features without checking for uniqueness. 

You're right, uniqueness is not currently enforced, I think it should be, though. Otherwise some later part of the pipeline might fail (like it did for me in the linked issue) and users have to go through a debugging process to figure out that the real cause is actually OneHotEncoder.

> We also add a little more overhead for uniqueness checking.

This is a valid point. How latency sensitive do you think get_feature_names_out is?

> I think we can reduce the scope of this PR by only including the callable support. We can follow up with a PR on uniqueness checking.

I can certainly do that if you think that there is another discussion needed around how we implement uniqueness check.

Reply:
> This is a valid point. How latency sensitive do you think get_feature_names_out is?

It will matter a bit when we want to support pandas output in the future, which would call `get_feature_names_out` in all transformers.

> I can certainly do that if you think that there is another discussion needed around how we implement uniqueness check.

It's worth opening an issue and discussing. The [SLEP007](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep007/proposal.html) does not say anything about uniqueness checking right now. I think there are valid arguments for uniqueness checking, but I would not want that discussion to block this PR. Supporting a callable is already a net improvement.

Reply:
> It's worth opening an issue and discussing. The [SLEP007](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep007/proposal.html) does not say anything about uniqueness checking right now. I think there are valid arguments for uniqueness checking, but I would not want that discussion to block this PR. Supporting a callable is already a net improvement.

Thanks for the additional context, I wasn't aware of SLEPs. If there's no policy around unique feature names, the question is really why ColumnTransformer checks that, but I guess that's a discussion for the original issue https://github.com/scikit-learn/scikit-learn/issues/22488.